staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go

```diff
@@ -104,6 +118,44 @@ func FinishRequest(ctx context.Context, fn ResultFunc) (runtime.Object, error) {
 	case result := <-resultCh:
 		return result.Return()
 	case <-ctx.Done():
-		return nil, errors.NewTimeoutError(fmt.Sprintf("request did not complete within requested timeout %s", ctx.Err()), 0)
+		// we are going to send a timeout response to the caller, but the asynchronous goroutine is
+		// still executing the ResultFunc function.
+		// kick off a go routine here to wait for the ResultFunc to return and then log some details.
+		defer func() {
+			go func() {
+				timedOutAt := time.Now()
+
+				var result *result
+				select {
+				case result = <-resultCh:
+				case <-time.After(postTimeoutWait):
+					// we will not wait forever, if we are here then we know that some goroutines
+					// are taking longer than postTimeoutWait.
+				}
+				postTimeoutLogger(timedOutAt, result)
+			}()
+		}()
+		return nil, errors.NewTimeoutError(fmt.Sprintf("request did not complete within requested timeout - %s", ctx.Err()), 0)
 	}
 }
+
+// logPostTimeoutResult logs a panic or an error from the result of a ResultFunc function
+// after the request had timed out.
+// timedOutAt is the time the request had been timed out
+func logPostTimeoutResult(timedOutAt time.Time, r *result) {
+	if r == nil {
+		// r == nil indicates that the child goroutine never returned a result.
+		klog.Errorf("FinishRequest: post-timeout activity, waited for %s, child goroutine has not returned yet", time.Since(timedOutAt))
+		return
+	}
+	if r.err == nil && r.reason == nil {
+		// a request returning with a success is not so interesting to us
+		return
+	}
+
+	// a non empty reason indicates that there was a panic.
+	metrics.RecordRequestPostTimeout("rest-handler", r.reason != nil)
+	err := fmt.Errorf("FinishRequest: post-timeout activity - time-elapsed: %s, panicd: %t, panic-reason: %v",
+		time.Since(timedOutAt), r.reason != nil, r.reason)
```

Comment:
what if `r.reason` is nil? will it panic?

Reply:
it should not because of `%v`

Reply:
actually, I have a test for this and here is the output:
```
E0413 09:51:23.299443  127665 finisher.go:167] FinishRequest: post-timeout activity - time-elapsed: 1.702µs, panicd: false, panic-reason: <nil>
```

Reply:
thanks, now I recall that it would dereference the value stored at the pointer but it won't use it